palace/fem/libceed/restriction.cpp

```diff
@@ -0,0 +1,298 @@
+                     bool is_range, Ceed ceed, CeedElemRestriction *restr)
+{
+  // Check for fespace -> restriction in hash table.
+  // The restriction for an interpolator range space is slightly different as
+  // the output is a primal vector instead of a dual vector, and lexicographic
+  // ordering is never used (no use of tensor-product basis).
+  // A palace::FiniteElementSpace can be checked for uniqueness so we can use this to reuse
+  // restrictions across different libCEED operators. For mixed meshes or multiple threads,
+  // the space elements are partitioned in a non-overlapping manner so we just need the
+  // index of the first element, and if it is a domain or boundary element, to determine the
+  // partition.
+  const FiniteElementSpace *restr_fespace =
+      dynamic_cast<const FiniteElementSpace *>(&fespace);
+  MFEM_VERIFY(restr_fespace, "ceed::InitRestriction requires a palace::FiniteElementSpace "
+                             "object for space comparisons!");
+  const mfem::FiniteElement &fe =
+      use_bdr ? *fespace.GetBE(indices[0]) : *fespace.GetFE(indices[0]);
+  const mfem::TensorBasisElement *tfe = dynamic_cast<const mfem::TensorBasisElement *>(&fe);
+  const bool vector = fe.GetRangeType() == mfem::FiniteElement::VECTOR;
+  mfem::Array<int> dofs;
+  mfem::DofTransformation dof_trans;
+  if (use_bdr)
+  {
+    fespace.GetBdrElementDofs(indices[0], dofs, dof_trans);
+  }
+  else
+  {
+    fespace.GetElementDofs(indices[0], dofs, dof_trans);
+  }
+  const bool has_dof_trans = dof_trans.GetDofTransformation() && !dof_trans.IsEmpty();
+  const bool unique_interp_restr =
+      (is_interp && tfe && tfe->GetDofMap().Size() > 0 && !vector);
+  const bool unique_interp_range_restr = (is_interp && is_range && has_dof_trans);
+  internal::RestrKey key(ceed, *restr_fespace, indices[0], use_bdr, unique_interp_restr,
+                         unique_interp_range_restr);
+
+  // Initialize or retrieve key values (avoid simultaneous search and write).
+  auto restr_itr = internal::restr_map.end();
+  PalacePragmaOmp(critical(InitRestriction))
+  {
+    restr_itr = internal::restr_map.find(key);
+  }
+  if (restr_itr == internal::restr_map.end())
+  {
+    const bool lexico = (tfe && tfe->GetDofMap().Size() > 0 && !vector && !is_interp);
+    if (lexico)
+    {
+      // Lexicographic ordering using dof_map.
+      InitLexicoRestr(fespace, indices, use_bdr, ceed, restr);
+    }
+    else
+    {
+      // Native ordering.
+      InitNativeRestr(fespace, indices, use_bdr, has_dof_trans, is_interp && is_range, ceed,
+                      restr);
+    }
+    PalacePragmaOmp(critical(InitRestriction))
+    {
+      internal::restr_map[key] = *restr;
+    }
+    // std::cout << "New element restriction (" << ceed << ", " << &fespace
+    //           << ", " << indices[0] << ", " << use_bdr
+    //           << ", " << unique_interp_restr
+    //           << ", " << unique_interp_range_restr << ")\n";
```

Comment:
`if constexpr(false)` for placing commented out code ensures it will still compile if/when we come to need it.

Reply:
I think it's fine to leave as is commented out, it's just one line.